llvm/lib/Transforms/InstCombine/InstructionCombining.cpp

```diff
@@ -4052,7 +4098,26 @@ Instruction *InstCombinerImpl::visitExtractValueInst(ExtractValueInst &EV) {
   if (Value *V = simplifyExtractValueInst(Agg, EV.getIndices(),
                                           SQ.getWithInstruction(&EV)))
     return replaceInstUsesWith(EV, V);
-
+  if (EV.getNumIndices() == 1 && EV.getIndices()[0] == 0) {
+    if (auto *FrexpCall = dyn_cast<CallInst>(Agg)) {
+      if (Function *F = FrexpCall->getCalledFunction()) {
+        if (F->getIntrinsicID() == Intrinsic::frexp) {
+          if (auto *SelInst =
+                  dyn_cast<SelectInst>(FrexpCall->getArgOperand(0))) {
+            if (isa<ConstantFP>(SelInst->getTrueValue()) ||
+                isa<ConstantFP>(SelInst->getFalseValue())) {
+              Builder.SetInsertPoint(&EV);
```

Comment:
Don't pre-check the conditions or adjust the insert point here. Everything belongs in the foldFrexpOfSelect implementation